### ISSUE TITLE
Class not found when using a class from a file which is not in the Standards folder.

### ISSUE THREAD

**Issue by Dakwamine** [squizlabs/PHP_CodeSniffer#1430]
I am trying to upgrade a standard, named *MyStandard* for the purpose of this issue, from phpcs 2 to 3. My standard's folder is located somewhere on my computer and is referenced in the `installed_paths` config. The standard is detected.

In *MyStandard*, some sniffs are using an abstract Sniff class which is also located in *MyStandard*. So for a file named *MySniff.php*:

```php
<?php

namespace PHP_CodeSniffer\Standards\MyStandard\Sniffs\CategoryName;

use PHP_CodeSniffer\Standards\MyStandard\Sniffs\CategoryName\SomeAbstractSniff;
use PHP_CodeSniffer\Files\File;

class MySniff extends SomeAbstractSniff
{ ... }
```

In this case, *SomeAbstractSniff.php* is located in the same folder as the file mentioned above. But when I am trying to run the standard with `phpcs --standard=MyStandard file.php`, there is an error:

```
PHP Fatal error:  Class 'PHP_CodeSniffer\Standards\MyStandard\Sniffs\CategoryName\SomeAbstractSniff' not found
```

What happens is that the autoloader (*autoload.php*) is not getting the file at the right location: it tries to load it from the *Standards* folder and not in *MyStandard*.

Am I forgetting something? Is it a bug of the autoloader, or do I have to implement one (or rely on requires) in my own standard's folder?

# Reproduction steps
1. Move the *Generic* standard folder outside the *Standards* folder, for example at: `C:\some\path`.
2. Reference the path in the *CodeSniffer.conf* file. Now it contains:
    > `'installed_paths' => 'C:\\some\\path`
3. Run `phpcs --standard=Generic file.php`:
    > `PHP Fatal error:  Class 'PHP_CodeSniffer\Standards\Generic\Sniffs\PHP\ForbiddenFunctionsSniff' not found in C:\some\path\Generic\Sniffs\PHP\DeprecatedFunctionsSniff.php on line 13`

# Environment
* Installed 3.0.0RC4 via composer: `composer global require squizlabs/php_codesniffer:3.0.0RC4`.
* Windows 7, Git Bash and cmd.exe.

# Workarounds
1. Place the standard's folder directly in the *Standards* folder.
2. Use:
    ```php
    require_once(dirname(__FILE__).DIRECTORY_SEPARATOR.'SomeAbstractSniff.php');
    ```

**Comment by gsherwood** [comment]
I haven't had time to look into this yet, but off the top of my head, I don't remember doing anything to help autoload files outside of the PHPCS root. If you were using composer to install your custom standard, the composer autoloader would kick in to do autoloading of those files. But as your standard is not a composer package, you'd need to require the non-sniff files manually.

**Comment by gsherwood** [comment]
I've taken a quick look and confirmed that there is nothing in the autoloader to help look for files outside the PHPCS root.

The fact that you've used the PHP_CodeSniffer namespace for your sniff files is what is causing the autoloader to go looking in the main `Standards` directory. If you used your own namespace, you'd just end up with the autoloader returning `false` and nothing being autoloaded. Either way, the autoloader for PHPCS is *only* looking for code that ships with PHPCS.

I'll take a look and see if it could also go looking into the installed paths, but it needs some more investigation. It's likely that PHPCS will have to modify it's autoloaders as new standard paths are discovered during a run, and/or the installed paths settings needs to be pre-loaded into the autoloader once the run begins.

**Comment by gsherwood** [comment]
I've pushed up a change that will allow autoloading to work inside loaded coding standards. In your case, it should work fine as long as you change your namespace.

You can't use the PHP_CodeSniffer namespace because PHPCS assumes you are looking for core files.

So instead of `PHP_CodeSniffer\Standards\MyStandard\Sniffs\CategoryName`, use `MyStandard\Sniffs\CategoryName` and autoloading should work.

Does that solve your problem?

**Comment by Dakwamine** [comment]
Sorry for the delay. Yes! It works wonderfully. Many thanks!

**Comment by gsherwood** [comment]
Good news. Thanks for testing that for me.

**Comment by Dakwamine** [comment]
Sorry for not asking before. Will this enhancement be included in the next RC/release?

**Comment by gsherwood** [comment]
> Sorry for not asking before. Will this enhancement be included in the next RC/release?

Yes. Just in the 3.x branch. Hopefully the next one in 3.0.0 stable.
